### PR TITLE
Paragraph segmentation

### PR DESCRIPTION
**Pull request by jonomon**

### CODE REVIEW

File: scripts/paragraph_segmentation_dcnn.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import random
+import os
+
+import matplotlib.pyplot as plt
+import matplotlib.patches as patches
+
+import mxnet as mx
+import numpy as np
+from skimage.draw import line_aa
+from skimage import transform as tf
```
Comment (ThomasDelteil) [inline scripts/paragraph_segmentation_dcnn.py]: please do not use `tf`, this is the common abbreviation for Tensorflow 😄

File: scripts/paragraph_segmentation_dcnn.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+def transform(data, label):
+    image = mx.nd.array(data).expand_dims(axis=2)
+    image = resize_short(image, int(800/3))
+    image = image.transpose([2, 0, 1])/255.
+    label = label[0].astype(np.float32)
+    return image, mx.nd.array(label)
+
+def augment_transform(data, label):
+    width_range = 0.05
```
Comment (ThomasDelteil) [inline scripts/paragraph_segmentation_dcnn.py]: add some comments to describe what the transforms are doing

File: scripts/paragraph_segmentation_dcnn.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            test_output_image = draw_box_on_image(pred_np, label_np, data_np)
+
+    train_loss = float(loss.asscalar())/len(train_data)
+    test_loss = float(test_loss.asscalar())/len(test_data)
+    
+    if e % print_every_n == 0 and e > 0:
+        print("Epoch {0}, train_loss {1:.6f}, val_loss {2:.6f}".format(
+            e, train_loss, test_loss))
+    
+    with SummaryWriter(logdir='./logs', verbose=False, flush_secs=5) as sw:
```
Comment (ThomasDelteil) [inline scripts/paragraph_segmentation_dcnn.py]: make the log dir parametrizable

File: scripts/paragraph_segmentation_dcnn.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    cnn.add(gluon.nn.Dense(4, activation='sigmoid'))
+
+    cnn.hybridize()
+    return cnn
+
+cnn = make_cnn()
+cnn.load_params("model_checkpoint/cnn300.params", ctx=ctx)
+
+iou_loss = IOU_loss()
+LEARNING_RATE = 0.0001
```
Comment (ThomasDelteil) [inline scripts/paragraph_segmentation_dcnn.py]: please have the parameters as argparse parameters to allow for future HPO.
at least:
LEARNING_RATE
BATCH_SIZES
epochs
log dir
checkpoint dir

data augmentation params?
cnn num layers?

File: scripts/paragraph_segmentation_dcnn_pretraining.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+from mxnet.image import resize_short
+from mxboard import SummaryWriter
+
+ctx = mx.gpu()
+mx.random.seed(1)
+
+from utils.iam_dataset import IAMDataset
+from utils.draw_box_on_image import draw_box_on_image
+
+MODEL_CHECKPOINT_FOLDER = "model_checkpoint"
```
Comment (ThomasDelteil) [inline scripts/paragraph_segmentation_dcnn_pretraining.py]: same comment as parametrizing the training script file using argparse

File: scripts/paragraph_segmentation_dcnn_pretraining.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    test_loss = nd.zeros(1, ctx)
+    tick = time.time()
+    acc = nd.zeros(1, ctx)
+    for i, (data, label) in enumerate(train_data):
+        data = data.as_in_context(ctx)
+        label = label.as_in_context(ctx)
+        
+        with autograd.record():
+            output = cnn(data)
+            loss_l2 = l2_loss(output, label)
```
Comment (ThomasDelteil) [inline scripts/paragraph_segmentation_dcnn_pretraining.py]: is this the only difference between the IOU loss and the pretraining script?
Please combine and use a parameter to switch the loss, and another one to point to a give checkedpoint model

File: utils/draw_box_on_image.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
+import numpy as np
+from skimage.draw import line_aa
+import matplotlib.pyplot as plt
+
+def draw_line(image, y1, x1, y2, x2, line_type):
```
Comment (ThomasDelteil) [inline utils/draw_box_on_image.py:5]: please add docstring

File: utils/draw_box_on_image.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    rr, cc, val = line_aa(y1, x1, y2, x2)
+    if line_type == "dotted":
+        rr = np.delete(rr, np.arange(0, rr.size, 5))
+        cc = np.delete(cc, np.arange(0, cc.size, 5))
+    image[rr, cc] = 0
+    return image
+    
+def draw_box(bounding_box, image, line_type):
```
Comment (ThomasDelteil) [inline utils/draw_box_on_image.py:42]: please add docstring

File: utils/iam_dataset.py [hunk 9]
```diff
@@ -258,7 +252,20 @@ def _get_bb_of_item(self, item, height, width):
         x2 = float(x2) / width
         y1 = float(y1) / height
         y2 = float(y2) / height
-        return [x1, y1, x2 - x1, y2 - y1]
+        bb = [x1, y1, x2 - x1, y2 - y1]
+
+        # Expand the bounding box by 3% to relax the boundaries
+        s = 0.03
```
Comment (ThomasDelteil) [inline utils/iam_dataset.py]: This shouldn't be here IMO it should be part of the transform, and could you make it parametric?

File: utils/iou_loss.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    Parameters
+    ----------
+    weight : float or None
+        Global scalar weight for loss.
+    batch_axis : int, default 0
+        The axis that represents mini-batch.
+
+
+    Inputs:
+        - **pred**: prediction tensor with arbitrary shape
```
Comment (ThomasDelteil) [inline utils/iou_loss.py]: please describe that the input should be [x, y, w, h]

File: utils/iou_loss.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+          sample_weight should have shape (64, 1).
+
+    Outputs:
+        - **loss**: IOU loss tensor with shape (batch_size,).
+    """
+
+    def __init__(self, weight=1., batch_axis=0, **kwargs):
+        super(IOU_loss, self).__init__(weight, batch_axis, **kwargs)
+        
+    def hybrid_forward(self, F, pred, label, sample_weight=None):
```
Comment (ThomasDelteil) [inline utils/iou_loss.py:33]: please add comments to guide the reader through the computations

File: scripts/paragraph_segmentation_dcnn.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        
+        with autograd.record():
+            output = cnn(data)
+            loss_iou = iou_loss(output, label)
+        loss_iou.backward()
+        loss += loss_iou.mean()
+        
+        trainer.step(data.shape[0])
+        if e % send_image_every_n == 0 and e > 0 and i == 0:
+            output = cnn(data)
```
Comment (ThomasDelteil) [inline scripts/paragraph_segmentation_dcnn.py]: unnecessary

File: scripts/paragraph_segmentation_dcnn.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+            data_np = data.asnumpy()
+            label_np = label.asnumpy()
+            pred_np = output.asnumpy()
+            train_output_image = draw_box_on_image(pred_np, label_np, data_np)
```
Comment (ThomasDelteil) [inline scripts/paragraph_segmentation_dcnn.py]: simplify and remove line 121->123

File: scripts/paragraph_segmentation_dcnn.py [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+LEARNING_RATE = 0.0001
+trainer = gluon.Trainer(cnn.collect_params(), 'adam', {'learning_rate': LEARNING_RATE, })
+
+epochs = 500
+print_every_n = 20
+send_image_every_n = 20
+save_every_n = 100
+
+for e in range(epochs):    
+    loss = nd.zeros(1, ctx)
```
Comment (ThomasDelteil) [inline scripts/paragraph_segmentation_dcnn.py]: make this a `run_epoch(net, dataloader, loss_function) -> loss` and simplify your loop

File: .gitignore [hunk 15]
```diff
@@ -2,4 +2,7 @@
 *~
 handwritting/*
 dataset/*
-credentials.json
\ No newline at end of file
+credentials.json
+logs/*
+nohup.out
+model_checkpoint/*
```
Comment (ThomasDelteil) [inline .gitignore:8]: Can't comment on the notebook as it is too large, will comment here:
`paragraph_segmentation_dcnn.ipynb`
=> make the training loop, evaluation loop dependent on a `run_epoch(net, dataloader, loss_function)` function and simplify your notebook by using it in each place instead of creating a new loop.

Comment (ThomasDelteil) [reply .gitignore:8]: `paragraph_segmentation_msers.ipynb` => good research notebook, though could have a bit more explanation than "doesn't generalize well" where all the above examples are example of perfect segmentation. One or two failure case would be nice to elaborate on that.

**Review by ThomasDelteil — CHANGES_REQUESTED** [review]
Great work :+1:, please refactor according to feedback and make the training script parametric. Thanks!

File: scripts/paragraph_segmentation_dcnn.py (hunk 12 above)
Comment (jonomon) [reply scripts/paragraph_segmentation_dcnn.py]: this line trainer.step(data.shape[0]) ?

File: paragraph_segmentation_dcnn.py [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+from utils.iam_dataset import IAMDataset
+from utils.iou_loss import IOU_loss
+from utils.draw_box_on_image import draw_box_on_image
+
+print_every_n = 20
+send_image_every_n = 20
+save_every_n = 100
+
+# pre-training: python paragraph_segmentation_dcnn.py -r 0.001 -e 301 -n cnn_mse.params
+# fine-tuning: python paragraph_segmentation_dcnn.py -r 0.0001 -l iou -e 150 -n cnn_iou.params -f cnn_iou.params
```
Comment (ThomasDelteil) [inline paragraph_segmentation_dcnn.py]: restore from `cnn_mse.params` no?

File: train_paragraph_segmentation.sh [hunk 17]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+python paragraph_segmentation_dcnn.py -r 0.001 -e 301 -n cnn_mse.params
+python paragraph_segmentation_dcnn.py -r 0.0001 -l iou -e 150 -n cnn_iou.params
```
Comment (ThomasDelteil) [inline train_paragraph_segmentation.sh]: loading checkpoint missing?

File: utils/draw_box_on_image.py [hunk 18]
```diff
@@ -3,6 +3,35 @@
...
+        Starting x position of the line
+
+    y2: int
+        Ending y position of the line
+
+    x2: int
+        Ending x position of the line
+
+    line_type: str
+        Options of the line_type is dotted or full.
```
Comment (ThomasDelteil) [inline utils/draw_box_on_image.py]: please list options in doc string

File: utils/draw_box_on_image.py [hunk 19]
```diff
@@ -11,6 +40,28 @@ def draw_line(image, y1, x1, y2, x2, line_type):
...
+    Parameters
+    ----------
+    bounding_box: [x, y, w, h]
+        The predicted bounding boxes in percentages
+
+    images: np.array
+        The correponding image.
+
+    line_type: str
+        Options of the line_type is dotted or full.
```
Comment (ThomasDelteil) [inline utils/draw_box_on_image.py]: please list valid options

File: utils/draw_box_on_image.py (hunk 18 above)
Comment (jonomon) [reply utils/draw_box_on_image.py]: the options are "dotted" or "full"

File: paragraph_segmentation_dcnn.py [hunk 20]
```diff
@@ -0,0 +1,218 @@
+import sys
+sys.path.append("..")
```
Comment (ThomasDelteil) [inline paragraph_segmentation_dcnn.py]: what is this?

Comment (jonomon) [reply paragraph_segmentation_dcnn.py]: A mistake

File: utils/draw_box_on_image.py (hunk 18 above)
Comment (ThomasDelteil) [reply utils/draw_box_on_image.py]: 👍 please add quotes in the doc string then. Or list like: `Type of line: ["dotted", "full"]`

File: paragraph_segmentation_dcnn.py [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+        if update_cnn:
+            loss_i.backward()
+            trainer.step(data.shape[0])
+
+        total_loss += loss_i.mean()
+        
+        if e % send_image_every_n == 0 and e > 0 and i == 0:
+            output_image = draw_box_on_image(output.asnumpy(), label.asnumpy(), data.asnumpy())
+            
+    epoch_loss = float(total_loss .asscalar())/len(dataloader)
```
Comment (ThomasDelteil) [inline paragraph_segmentation_dcnn.py]: extra space here

**Review by ThomasDelteil — APPROVED** [review]

**Comment by ThomasDelteil** [comment]
@jonomon sounds good! Please fix the conflicts and merge.
